releng: Enable elevated privileges for Release Managers on SIG Release GCP projects

justaugustus opened the pull request:
- Add groups to delegate IAM access to Release Managers

  - `k8s-infra-release-admins@`: Partial Admin
  - `k8s-infra-release-editors@`: Editor
  - `k8s-infra-release-viewers@`: Viewer

- Add ensure-release-projects to grant rights to Release Managers
  Establishes rights to GCS, GCB, and KMS for Release Managers

Required for prototyping build/stage/release process on k8s-infra (https://github.com/kubernetes/release/issues/911).

/assign @dims @thockin @cblecker 
cc: @kubernetes/sig-release-admins @kubernetes/release-engineering 
/sig release
/area release-eng

cblecker commented:
/hold
Explicit hold for review.

hoegaarden commented on infra/gcp/ensure-release-projects.sh:
```diff
@@ -0,0 +1,140 @@
+    color 3 "Configuring: ${REPO}"
+
+    # The GCP project name.
+    PROJECT="${REPO}"
```
Can we maybe just use `PROJECT` instead of `REPO`? As far as I can tell, we never use `REPO` further down, so can we just do `for PROJECT ...`?

thockin replied:
yes, the loop can be `for PROJECT` - "repo" is a holdover from the script he copied, I assume :)

hoegaarden commented on infra/gcp/lib.sh:
```diff
@@ -287,6 +287,23 @@ function empower_group_for_gcb() {
+# $1: The GCP project
+# $2: The group email
+function empower_group_for_kms() {
+    if [ $# -lt 2 -o -z "$1" -o -z "$2" ]; then
```
Should we make that stricter by doing something like the below?
```bash
if [ $# -ne 2 -o -z "$1" -o -z "$2" ]; then
```

I saw, that other funcs here allow additional arguments, but that might just be a c&p "issue" from cases where additional optional arguments are allowed (e.g. `empower_group_to_gcr` or `empower_gcr_admins`).

thockin replied:
this is probably a good idea, but I'm OK to let this PR go in and fix it up across the board.

thockin commented on infra/gcp/ensure-release-projects.sh:
```diff
@@ -0,0 +1,140 @@
+    echo "usage: $0 [repo...]" > /dev/stderr
+    echo "example:" > /dev/stderr
+    echo "  $0 # do all staging repos" > /dev/stderr
+    echo "  $0 coredns # just do one" > /dev/stderr
```
change "coredns" to a real (or realer-looking) example?

thockin commented on infra/gcp/ensure-release-projects.sh:
```diff
@@ -0,0 +1,140 @@
+    set -- "${PROJECTS[@]}"
+fi
+
+ADMINS="release-managers-admins@kubernetes.io"
```
Casually looking at this, the distinctions are not clear.  Why are release-managers just viewers?  Do we need to rename groups?  Can this be solved with docs here?  Why do we need 3 distinct groups?

justaugustus replied:
Here's what each group would do:
- `release-managers@`: Includes all Release Managers (including Release Manager Associates), so this needs to be least-privilege (allow Associates to view, but not create, edit, delete)
- `release-managers-private@`: Only includes Patch Release Team, Branch Managers, Build Admins, SIG Chairs (the set of users we want working on GCP, so we need edit here)
- `release-managers-admins@`: Only includes SIG Release Chairs and grants additional access to administer KMS (limited set of users, security event mitigation, key mgmt (store GitHub tokens, deb/rpm signing key))

cblecker replied:
We should make distinctions here between "Groups that provide IAM roles" and "E-mail distribution lists". I would be concerned with someone getting access to the release artifacts simply because they become a SIG chair and need access to an e-mail list.

thockin replied:
I agree with @cblecker - there are the social groups (email) and the permissions groups, and I am not sure they are the same.   Should being a SIG-chair automatically grant you access to GCP?

thockin commented on infra/gcp/ensure-release-projects.sh:
```diff
@@ -0,0 +1,140 @@
+
+# NB: Please keep this sorted.
+PROJECTS=(
+    k8s-staging-release-test
```
currently these are covered by the other ensure-* scripts.  Should we a) drop them from those, b) add this logic to those, c) refactor to make the overlap clear?

justaugustus commented:
@thockin --
> Mostly I question the overall factoring - these don't feel like "staging" and "prod" projects (in the sense that we have scripts that exist) any more.

Yeah, totally agreed. This felt awkward to do, as it felt like shimming something into a model it doesn't really fit. 
Each of the SIG Release GCP projects are essentially near-prod/prod, as various areas of the project depend on the artifacts:
- `k8s-staging-release-test/ci`: Prow/intrepid users testing nightlies
- `k8s-staging-release-test/stage`: Validated prior to promotion (but wider area of effect than other staging projects)
- `k8s-release-test-prod/release`: All Kubernetes users

I think this projects would be more accurately named:
- `k8s-release-staging`
- `k8s-release-prod`

> Maybe they need to be totally distinct (removed from other scripts) and even renamed?
> currently these are covered by the other ensure-* scripts. Should we a) drop them from those, b) add this logic to those, c) refactor to make the overlap clear?

In an ideal world/future refactor, perhaps there could be one script, `configure-projects.sh`?
It would include three arrays, `STAGING_PROJECTS`, `PROD_PROJECTS`, and `RELEASE_PROJECTS` and maybe ensure behavior/project config/principal access on a set of `case` statements?

The overlap here is primarily because of copy/pasta.

I think the future should also include yaml config for custom roles, instead of adhoc IAM attaches.

justaugustus commented:
(Just also noting that I went the copy/pasta route instead of refactoring because it wouldn't interfere with the other projects.)

justaugustus commented:
(@hoegaarden -- I'll come back and address your review once we land on an approach.)

cblecker commented on groups/groups.yaml:
```diff
@@ -1026,7 +1051,7 @@ groups:
       WhoCanModifyMembers: "OWNERS_AND_MANAGERS"
       WhoCanModerateMembers: "OWNERS_AND_MANAGERS"
-      ReconcileMembers: "false"
+      ReconcileMembers: "true"
```
This will remove 30 members from the list. Is this intended?

justaugustus replied:
Dropped this commit.

justaugustus commented:
@cblecker --
> I've mentioned this before in other contexts, but tying privileges to "chairpersonship" of a SIG/WG is an anti-pattern we should try to avoid. These privileges should be tied to A) a unique role defined by the SIG (like "Release Admins" or whatever you want to call it) or B) ownership of a subproject, or even C) a list of people who are just "keyholders".

Yep, yep. I think the last time we chatted about it was here: https://github.com/kubernetes/org/pull/1205#discussion_r326947353.
SIG Chairs are also Release Engineering subproject owners, so I occasionally interchange the two in documentation. I've updated the verbiage there.

> We should make distinctions here between "Groups that provide IAM roles" and "E-mail distribution lists". I would be concerned with someone getting access to the release artifacts simply because they become a SIG chair and need access to an e-mail list.

Combining the two (communication and access) is exactly what we wanted to do for the Release Manager lists. We also moved to kubernetes.io for a more "official" address and more importantly, to get the benefit of being able to manage this access via config/PR.

More details can be found here: https://github.com/kubernetes/sig-release/issues/732

thockin commented on groups/groups.yaml:
```diff
@@ -1002,6 +1002,31 @@ groups:
+    owners:
+      - caselim@gmail.com
+      - cmiles@pivotal.io
+      - saugustus@vmware.com
```
why are we adding two accounts for Stephen and Tim?  This means twice as many credentials that could leak.  I'd like to make a policy that we explicitly NOT do this.

thockin commented on infra/gcp/ensure-release-projects.sh:
```diff
@@ -0,0 +1,140 @@
+function usage() {
+    echo "usage: $0 [repo...]" > /dev/stderr
+    echo "example:" > /dev/stderr
+    echo "  $0 # do all staging repos" > /dev/stderr
```
s/staging repos/release projects

thockin commented on infra/gcp/ensure-release-projects.sh:
```diff
@@ -0,0 +1,140 @@
+    echo "  $0 # do all staging repos" > /dev/stderr
+    echo "  $0 coredns # just do one" > /dev/stderr
+    echo > /dev/stderr
+}
```
it serves as docs, at least :)

thockin commented on infra/gcp/ensure-release-projects.sh:
```diff
@@ -0,0 +1,140 @@
+WRITERS="release-managers-private@kubernetes.io"
+VIEWERS="release-managers@kubernetes.io"
+
+for REPO; do
```
Once I learned to embrace "$@" I started to prefer the simpler syntax, and this PR follows wstablished patterns

thockin commented on infra/gcp/ensure-release-projects.sh:
```diff
@@ -0,0 +1,140 @@
+    PROJECT="${REPO}"
+
+    # The names of the buckets
+    STAGING_BUCKET="gs://${PROJECT}" # used by humans
```
This is the piece I want to discuss most - Do we want these projects "governed" by 2 ensure scripts?  This overlaps directly with the other scripts, which means it will drift.

There's a particular set of semantics that we ascribe to a "staging project" and a "prod project".  You are adding on to that, clearly, but it's not clear if you are changing more than that.  In other words - should we make a clean break and remove these release projects from the other scripts?  Or should we build on those?

For example, this script could do something like:


```
# First ensure that the project is a normal staging project
"${ROOT}/ensure-staging-project.sh" release-test
# Now apply customizations for release staging
```

If it helps, we can refactor the existing scripts to make this more obvious.  I knew this was likely to come up, so we it should not be hard - we left room.

thockin commented:
Thanks!

/lgtm
/approve

justaugustus commented:
Awesome. Thanks y'all!
/hold cancel

cblecker commented:
/approve

cblecker commented:
Groups have been created.
